File: LeetCode/14 M7 - Leetcode 18 - 4Sum/hashing/18-external.js
// https://leetcode.com/problems/4sum/discuss/1341602/Python-or-Javascript-or-HashMap


/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    
    let hashMap = {};
    let output = [];
    let uniqueMap = {};
    
    for (let i = 0; i < nums.length; i++){
        for (let j = i+1; j < nums.length; j++){
            
            if (hashMap[nums[i]+nums[j]] ){
                hashMap[nums[i]+nums[j]].push([i, j, nums[i], nums[j]]);
            } else {
                hashMap[nums[i]+nums[j]] = [[i, j, nums[i], nums[j]]];
            }
        }
    }
    for (let i = 0; i < nums.length; i++){
        for (let j = i+1; j < nums.length; j++){
            let complement = target - (nums[i]+nums[j]);
            
            if(hashMap[complement]){
                // Now I went to see if
                for (let k = 0; k < hashMap[complement].length; k++){
                    
                    if (i !=  hashMap[complement][k][0] && i != hashMap[complement][k][1] && j != hashMap[complement][k][0] && j != hashMap[complement][k][1]){
                        
                        temp = [nums[i], nums[j], hashMap[complement][k][2], hashMap[complement][k][3]];
                        temp.sort();
                        
                        tempString = temp.toString();
                
                        if (!uniqueMap[tempString]){
                            output.push(temp);
                            uniqueMap[tempString] = 1 ;
                        }
                      
                        
                        
                    }
            }
        }
    }  
    }
    return output;
};

File: LeetCode/14 M7 - Leetcode 18 - 4Sum/2-pointers/18-external.py

def fourSum(nums, target):
    N = len(nums)
    if N < 4:
        return []
    res = []
    nums = sorted(nums)
    i = 0
    for i in range(N-3):
        if 0 < i < N-3 and nums[i] == nums[i-1]:
            continue        
        for j in range(i+1,N-2):
            if i+1 < j < N-2 and nums[j] == nums[j-1]:
                continue
            k = j+1
            l = N-1
            while k < l:
                summ = nums[i]+nums[j]+nums[k]+nums[l]
                if summ == target:
                    res.append([nums[i],nums[j],nums[k],nums[l]])
                    while k < N-1 and nums[k] == nums[k+1]:
                        k += 1
                    while l > 0 and nums[l] == nums[l-1]:
                        l -= 1
                    k += 1
                    l -= 1
                elif summ < target:
                    k += 1
                else:
                    l -= 1
    return res
print(fourSum([1,0,-1,0,-2,2], 0))